### ISSUE TITLE
[llvm-objcopy] seems to fail to add a gnu-debuglink on Go ld-output binaries on ppc64le

### ISSUE THREAD

**Issue by nekopsykose** [llvm/llvm-project#82170]
(and maybe aarch64?)  

for a start, i built https://github.com/yggdrasil-network/yggdrasil-go :  

```console
$ mkdir out
$ go build -o out/ ./cmd/...
```

then, i simulated splitting debuginfo and adding back a debuglink:  

```
$ llvm-objcopy --only-keep-debug out/yggdrasilctl x
$ llvm-objcopy --add-gnu-debuglink="$PWD"/x out/yggdrasilctl
$ llvm-objdump -s -j .gnu_debuglink out/yggdrasilctl

out/yggdrasilctl:	file format elf64-powerpcle
llvm-objdump: warning: section '.gnu_debuglink' mentioned in a -j/--section option, but not found in any input file
$ llvm-readelf -a out/yggdrasilctl >/dev/null
llvm-readelf: warning: 'out/yggdrasilctl': SHT_STRTAB string table section [index 9] is non-null terminated
llvm-readelf: error: 'out/yggdrasilctl': SHT_STRTAB string table section [index 9] is non-null terminated
```

seemingly, something exploded here. but, using ld.lld instead (/usr/bin/ld is lld, and external linkmode makes go not use its own linker and invoke ld normally):  

```
go build -o out/ -ldflags="-linkmode external" ./cmd/...
```

then works fine:  

```
$ llvm-objcopy --only-keep-debug out/yggdrasilctl x
$ llvm-objcopy --add-gnu-debuglink="$PWD"/x out/yggdrasilctl
$ llvm-objdump -s -j .gnu_debuglink out/yggdrasilctl

out/yggdrasilctl:	file format elf64-powerpcle
Contents of section .gnu_debuglink:
 0000 78000000 5b3c64ed                    x...[<d.
```

so, i am not sure if:  

- the go linker is broken on these architectures in some way and makes an invalid output (?)  
- llvm objcopy is not handling something here from the output. note the errors reported in llvm-readelf are not reported right after the binary is made- they only show up after running objcopy

for reference, the `yggdrasilctl` ppc64le binaries can be found here: https://img.ayaya.dev/3jO9b8kXBdbx
 (a .tar.zst). those are the go build outputs before any objcopy.

**Comment by jh7370** [comment]
Just to confirm, when you say:

> note the errors reported in llvm-readelf are not reported right after the binary is made

You've run llvm-readelf after the binary is made before llvm-objcopy is executed and can dump the section headers fine?

I'm not in a position currently to test out your binaries fully, but it might be worth using `llvm-readelf -x 9` to dump the supposedly invalid `SHT_STRTAB` section. This should produce a hex dump of the section and you can use that to see what might be going on (is there a missing null terminator, does the output look completely bogus etc). I wouldn't expect there to be a bug in llvm-objcopy regarding adding the section itself, because it uses the same logic as adding any other section. However, it's possible the input has some unexpected format that llvm-objcopy then gets confused by, does its best to work with, and then spits out garbage (i.e. garbage in garbage out).

**Comment by jh7370** [comment]
Which version of LLVM are you using? Have you tried with the current head of main? I have this vague memory that a similar issue may have been resolved, but I could be misremembering or it could be a different issue.

**Comment by nekopsykose** [comment]
> Which version of LLVM are you using?  

17.0.6

> Have you tried with the current head of main?   

nope. i guess i can try that ~soonish, i'll post when i do

**Comment by jh7370** [comment]
Okay, I was able to get GNU readelf to dump the hex output of the section header string table and this confirmed that the output has been messed up somehow (the table indeed doesn't have a null terminator). Output:
```
james@james-VirtualBox:~/Work$ readelf -x 9 rep/test
readelf: Error: no .dynamic section in the dynamic segment

Hex dump of section '':
  0x00000000 002e7465 7874002e 6e6f7074 72646174 ..text..noptrdat
  0x00000010 61002e64 61746100 2e627373 002e6e6f a..data..bss..no
  0x00000020 70747262 7373002e 676f2e66 757a7a63 ptrbss..go.fuzzc
  0x00000030 6e747273 002e676f 2e627569 6c64696e ntrs..go.buildin
  0x00000040 666f002e 74627373 002e6e6f 74652e67 fo..tbss..note.g
  0x00000050 6f2e6275 696c6469 64002e65 6c666461 o.buildid..elfda
  0x00000060 7461002e 726f6461 7461002e 74797065 ta..rodata..type
  0x00000070 6c696e6b 002e6974 61626c69 6e6b002e link..itablink..
  0x00000080 676f7379 6d746162 002e676f 70636c6e gosymtab..gopcln
  0x00000090 74616200 2e73796d 74616200 2e737472 tab..symtab..str
  0x000000a0 74616200 2e646562 75675f61 62627265 tab..debug_abbre
  0x000000b0 76002e64 65627567 5f667261 6d65002e v..debug_frame..
  0x000000c0 64656275 675f696e 666f002e 64656275 debug_info..debu
  0x000000d0 675f6c6f 63002e64 65627567 5f6c696e g_loc..debug_lin
  0x000000e0 65002e64 65627567 5f676462 5f736372 e..debug_gdb_scr
  0x000000f0 69707473 002e6465 6275675f 72616e67 ipts..debug_rang
  0x00000100 6573002e 73687374 72746162 002e696e es..shstrtab..in
  0x00000110 74657270 002e6861 7368002e 676f7400 terp..hash..got.
  0x00000120 2e676c69 6e6b002e 676f742e 706c7400 .glink..got.plt.
  0x00000130 2e64796e 616d6963 002e6479 6e73796d .dynamic..dynsym
  0x00000140 002e6479 6e737472 002e7265 6c61002e ..dynstr..rela..
  0x00000150 72656c61 2e706c74 002e706c 74002e67 rela.plt..plt..g
  0x00000160 6e752e76 65727369 6f6e002e 67       nu.version..g
```
(the error should have no impact on the actual behaviour). As can be seen, the last byte of the output is not a null terminator. It looks as though llvm-objcopy has set the size of the section name string table section incorrectly for some reason (note that it's actually shrunk in size, although the contents appear to be identical, for what can be seen). I'll prod around with the code a little more to see if I can spot any obvious culprits, but I doubt I'll spot something easily, and I'm not set up properly to debug things.

**Comment by jh7370** [comment]
Forgot to mention that the above output is using LLVM 14, which is what I have easily available to test with. @nekopsykose, can you please confirm that you see equivalent if you use GNU readelf to dump the section as described?

**Comment by nekopsykose** [comment]
yep:  

```
$ powerpc64le-chimera-linux-musl-greadelf -x 9 test

Hex dump of section '.shstrtab':
  0x00000000 002e7465 7874002e 6e6f7074 72646174 ..text..noptrdat
  0x00000010 61002e64 61746100 2e627373 002e6e6f a..data..bss..no
  0x00000020 70747262 7373002e 676f2e66 757a7a63 ptrbss..go.fuzzc
  0x00000030 6e747273 002e676f 2e627569 6c64696e ntrs..go.buildin
  0x00000040 666f002e 74627373 002e6e6f 74652e67 fo..tbss..note.g
  0x00000050 6f2e6275 696c6469 64002e65 6c666461 o.buildid..elfda
  0x00000060 7461002e 726f6461 7461002e 74797065 ta..rodata..type
  0x00000070 6c696e6b 002e6974 61626c69 6e6b002e link..itablink..
  0x00000080 676f7379 6d746162 002e676f 70636c6e gosymtab..gopcln
  0x00000090 74616200 2e73796d 74616200 2e737472 tab..symtab..str
  0x000000a0 74616200 2e646562 75675f61 62627265 tab..debug_abbre
  0x000000b0 76002e64 65627567 5f667261 6d65002e v..debug_frame..
  0x000000c0 64656275 675f696e 666f002e 64656275 debug_info..debu
  0x000000d0 675f6c6f 63002e64 65627567 5f6c696e g_loc..debug_lin
  0x000000e0 65002e64 65627567 5f676462 5f736372 e..debug_gdb_scr
  0x000000f0 69707473 002e6465 6275675f 72616e67 ipts..debug_rang
  0x00000100 6573002e 73687374 72746162 002e696e es..shstrtab..in
  0x00000110 74657270 002e6861 7368002e 676f7400 terp..hash..got.
  0x00000120 2e676c69 6e6b002e 676f742e 706c7400 .glink..got.plt.
  0x00000130 2e64796e 616d6963 002e6479 6e73796d .dynamic..dynsym
  0x00000140 002e6479 6e737472 002e7265 6c61002e ..dynstr..rela..
  0x00000150 72656c61 2e706c74 002e706c 74002e67 rela.plt..plt..g
  0x00000160 6e752e76 65727369 6f6e002e 676e752e nu.version..gnu.
  0x00000170 76657273 696f6e5f 7200              version_r.

$ llvm-objcopy --only-keep-debug test x
$ llvm-objcopy --add-gnu-debuglink="$PWD"/x test
$ powerpc64le-chimera-linux-musl-greadelf -x 9 test
readelf: Error: no .dynamic section in the dynamic segment

Hex dump of section '':
  0x00000000 002e7465 7874002e 6e6f7074 72646174 ..text..noptrdat
  0x00000010 61002e64 61746100 2e627373 002e6e6f a..data..bss..no
  0x00000020 70747262 7373002e 676f2e66 757a7a63 ptrbss..go.fuzzc
  0x00000030 6e747273 002e676f 2e627569 6c64696e ntrs..go.buildin
  0x00000040 666f002e 74627373 002e6e6f 74652e67 fo..tbss..note.g
  0x00000050 6f2e6275 696c6469 64002e65 6c666461 o.buildid..elfda
  0x00000060 7461002e 726f6461 7461002e 74797065 ta..rodata..type
  0x00000070 6c696e6b 002e6974 61626c69 6e6b002e link..itablink..
  0x00000080 676f7379 6d746162 002e676f 70636c6e gosymtab..gopcln
  0x00000090 74616200 2e73796d 74616200 2e737472 tab..symtab..str
  0x000000a0 74616200 2e646562 75675f61 62627265 tab..debug_abbre
  0x000000b0 76002e64 65627567 5f667261 6d65002e v..debug_frame..
  0x000000c0 64656275 675f696e 666f002e 64656275 debug_info..debu
  0x000000d0 675f6c6f 63002e64 65627567 5f6c696e g_loc..debug_lin
  0x000000e0 65002e64 65627567 5f676462 5f736372 e..debug_gdb_scr
  0x000000f0 69707473 002e6465 6275675f 72616e67 ipts..debug_rang
  0x00000100 6573002e 73687374 72746162 002e696e es..shstrtab..in
  0x00000110 74657270 002e6861 7368002e 676f7400 terp..hash..got.
  0x00000120 2e676c69 6e6b002e 676f742e 706c7400 .glink..got.plt.
  0x00000130 2e64796e 616d6963 002e6479 6e73796d .dynamic..dynsym
  0x00000140 002e6479 6e737472 002e7265 6c61002e ..dynstr..rela..
  0x00000150 72656c61 2e706c74 002e706c 74002e67 rela.plt..plt..g
  0x00000160 6e752e76 65727369 6f6e002e 67       nu.version..g
```

**Comment by jh7370** [comment]
@MaskRay, do you agree with my thought process in the previous post? If so, I think it's probably reasonable to add the good first issue tag to this ticket, since it should be easy enough for somebody to add an error. WDYT?

@nekopsykose, do you know why the .shstrtab section is placed in a segment by go's linker? Perhaps this needs reporting to them?

**Comment by nekopsykose** [comment]
> do you know why the .shstrtab section is placed in a segment by go's linker?  

i don't :) know nothing about linkers in this detail..  

> Perhaps this needs reporting to them?  

most definitely. i can do it when i find some time, though it would be nice if someone spied me to forward this with more technical detail than me just copypasting/linking this thread.

> Note that GNU objcopy (v2.38 at least) rejects the input file outright.  

hm. i did forget to test this on gnu binutils at all (oops..), 2.41 seems to copy it ok (and then reading it still works after). but this is a moot point- if this is invalid elf then it's invalid elf, no point in caring much more past that if it works.

agree that adding an explicit error sounds good if there's no other hangups :)

**Comment by jh7370** [comment]
> > do you know why the .shstrtab section is placed in a segment by go's linker?
> 
> i don't :) know nothing about linkers in this detail..

Fair enough. It may be that the go linker has a specific reason for putting it in a loadable segment, but I don't know anything about it, so I can't comment there - this is where we'd need their developers to feed back.

> > Perhaps this needs reporting to them?
> 
> most definitely. i can do it when i find some time, though it would be nice if someone sniped me to forward this with more technical detail than me just copypasting/linking this thread.

I think it's probably best we wait for @MaskRay to see if he has any input. I don't want to be misleading in my claims. In particular, I note that it doesn't _explicitly_ say that non-alloc sections cannot be in PT_LOAD, but it's implied (sections without SHF_ALLOC don't take up memory space, whilst PT_LOAD segments take up memory space by definition).

> > Note that GNU objcopy (v2.38 at least) rejects the input file outright.
> 
> hm. i did forget to test this on gnu binutils at all (oops..), 2.41 seems to copy it ok (and then reading it still works after). but this is a moot point- if this is invalid elf then it's invalid elf, no point in caring much more past that if it works.

Interesting. I'd be interested in knowing what's changed (or maybe it's something more mundane about my build of GNU objcopy that can't handle the file, like not having support for a specific platform).

**Comment by nekopsykose** [comment]
> Interesting. I'd be interested in knowing what's changed (or maybe it's something more mundane about my build of GNU objcopy that can't handle the file, like not having support for a specific platform).  

if it's like `gobjcopy: Unable to recognise the format of the input file `yggdrasilctl-go-ld'` then it would be the latter, yeah. check for `elf64-powerpcle` in `--help`

**Comment by jh7370** [comment]
> > Interesting. I'd be interested in knowing what's changed (or maybe it's something more mundane about my build of GNU objcopy that can't handle the file, like not having support for a specific platform).
> 
> if it's like `` gobjcopy: Unable to recognise the format of the input file `yggdrasilctl-go-ld' `` then it would be the latter, yeah. check for `elf64-powerpcle` in `--help`

Ah yes, that's it.

So when GNU objcopy adds a section to the file, what happens to the .shstrtab section? Does it remain within the PT_LOAD segment? What if there are no options specified (so a straight "copy")?

**Comment by nekopsykose** [comment]
> What if there are no options specified (so a straight "copy")?  

that moves the shstrtab to the end:  

```
There are 37 section headers, starting at offset 0x82b638:

Section Headers:
  [Nr] Name              Type            Address          Off    Size   ES Flg Lk Inf Al
  [ 0]                   NULL            0000000000000000 000000 000000 00      0   0  0
  [ 1] .text             PROGBITS        0000000000011000 001000 297b24 00  AX  0   0 32
  [ 2] .glink            PROGBITS        00000000002a8b40 298b40 0000c4 00  AX  0   0  4
  [ 3] .rodata           PROGBITS        00000000002b0000 2a0000 11943c 00   A  0   0 32
  [ 4] .gnu.version_r    VERNEED         00000000003c943c 3b943c 000000 00   A 10   0  1
  [ 5] .dynsym           DYNSYM          00000000003c9c40 3b9c40 000390 18   A 10   1  8
  [ 6] .rela             RELA            00000000003c9440 3b9440 000048 18   A  5   0  8
  [ 7] .rela.plt         RELA            00000000003c9488 3b9488 000318 18  AI  5  18  8
  [ 8] .gnu.version      VERSYM          00000000003c97a0 3b97a0 00004c 02   A  5   0 32
  [ 9] .hash             HASH            00000000003c9800 3b9800 0000bc 04   A  5   0  8
  [10] .dynstr           STRTAB          00000000003c9a40 3b9a40 0001e5 00   A  0   0  1
  [11] .typelink         PROGBITS        00000000003c9fe0 3b9fe0 001b34 00   A  0   0 32
  [12] .itablink         PROGBITS        00000000003cbb20 3bbb20 000a18 00   A  0   0 32
  [13] .gosymtab         PROGBITS        00000000003cc538 3bc538 000000 00   A  0   0  1
  [14] .gopclntab        PROGBITS        00000000003cc540 3bc540 17d070 00   A  0   0 32
  [15] .go.buildinfo     PROGBITS        0000000000550000 540000 0008c0 00  WA  0   0 16
  [16] .got.plt          PROGBITS        00000000005508c0 5408c0 000000 00  WA  0   0  1
  [17] .dynamic          DYNAMIC         00000000005508c0 5408c0 000110 10  WA 10   0  8
  [18] .plt              NOBITS          00000000005509e0 5409d0 000118 08  WA  0   0  8
  [19] .got              PROGBITS        0000000000550af8 540af8 000028 00  WA  0   0  8
  [20] .noptrdata        PROGBITS        0000000000550b20 540b20 03a5d0 00  WA  0   0 32
  [21] .data             PROGBITS        000000000058b100 57b100 00e5b0 00  WA  0   0 32
  [22] .bss              NOBITS          00000000005996c0 5896b0 036618 00  WA  0   0 32
  [23] .noptrbss         NOBITS          00000000005cfce0 5896b0 006e50 00  WA  0   0 32
  [24] .tbss             NOBITS          0000000000000000 5896b0 000010 00 WAT  0   0  8
  [25] .debug_abbrev     PROGBITS        0000000000000000 5896b0 000135 00   C  0   0  1
  [26] .debug_line       PROGBITS        0000000000000000 5897e5 06a59e 00   C  0   0  1
  [27] .debug_frame      PROGBITS        0000000000000000 5f3d83 018767 00   C  0   0  1
  [28] .debug_gdb_scripts PROGBITS       0000000000000000 60c4ea 000028 00      0   0  1
  [29] .debug_info       PROGBITS        0000000000000000 60c512 0eae9a 00   C  0   0  1
  [30] .debug_loc        PROGBITS        0000000000000000 6f73ac 09a8a0 00   C  0   0  1
  [31] .debug_ranges     PROGBITS        0000000000000000 791c4c 02b35f 00   C  0   0  1
  [32] .interp           PROGBITS        0000000000010fe2 000fe2 00001e 00   A  0   0  1
  [33] .note.go.buildid  NOTE            0000000000010f7c 000f7c 000064 00   A  0   0  4
  [34] .symtab           SYMTAB          0000000000000000 7bcfb0 031188 18     35 586  8
  [35] .strtab           STRTAB          0000000000000000 7ee138 03d39e 00      0   0  1
  [36] .shstrtab         STRTAB          0000000000000000 82b4d6 00015e 00      0   0  1
```

which seems to put it outside the LOAD (or well, i assume the load is delineated by this):  

```
Program Headers:
  Type           Offset   VirtAddr           PhysAddr           FileSiz  MemSiz   Flg Align
  PHDR           0x000040 0x0000000000010040 0x0000000000010040 0x0001f8 0x0001f8 R   0x10000
  INTERP         0x000fe2 0x0000000000010fe2 0x0000000000010fe2 0x00001e 0x00001e R   0x1
      [Requesting program interpreter: /lib/ld-musl-powerpc64le.so.1]
  NOTE           0x000f7c 0x0000000000010f7c 0x0000000000010f7c 0x000064 0x000064 R   0x4
  LOAD           0x000000 0x0000000000010000 0x0000000000010000 0x298c04 0x298c04 R E 0x10000
  LOAD           0x2a0000 0x00000000002b0000 0x00000000002b0000 0x2995b0 0x2995b0 R   0x10000
  LOAD           0x540000 0x0000000000550000 0x0000000000550000 0x0496b0 0x086b30 RW  0x10000
  DYNAMIC        0x5408c0 0x00000000005508c0 0x00000000005508c0 0x000110 0x000110 RW  0x8
  TLS            0x5896b0 0x0000000000000000 0x0000000000000000 0x000000 0x000010 R   0x8
  GNU_STACK      0x000000 0x0000000000000000 0x0000000000000000 0x000000 0x000000 RW  0x8
```

so, it seems that gnu objcopy 'fixes' the binary automatically. the same kind of result happens when adding the debuglink too.

using that 'fixed' binary (from pure copy only) after with llvm-objcopy exhibits no issues. fun :D

**Comment by jh7370** [comment]
Thanks, that's what I expected might happen. We could implement this in llvm-objcopy, which would essentially mean treating the section header string table as not being in a segment, even if it seems to be. This would leave the old section header names as ghost data in the segment (because segment contents are generally copied verbatim to the output, unless the user explicitly removes any contents), which seems reasonable, but depending on why the go linker adds it to the loadable segment in the first place might invalidate assumptions by the go tools.

**Comment by nekopsykose** [comment]
this seems to have been fixed in the go toolchain in `1.22`. since the issue was shstrtab, i had a quick search and found https://github.com/golang/go/issues/62600 which got fixed

so this is fixed on go toolchain side not generating malformed elfs :) sorry for only checking so late
